docs: glossary for naming conventions (#32)
rsksmart/rif-marketplace-storage-pinning: README.md

# RIF Marketplace Storage Pinning

```
npm i @rsksmart/rif-marketplace-storage-pinning
```

## PinningManager contract

### Message
The `StorageProvider` can emit an arbitrary message in the function calls `setStorageOffer` and `emitMessage`. 
The message is is initially intended to allow the `StorageProvider` to communicate his `nodeID`, but may be used for other purposes later.

The client should be instructed to enforce the following semantics:
- If `message` starts with `0x01`, then the following bits contain an arbitrary length `nodeID`.

## Glossary

 - Actors
    - **Consumer** - actor that wants his files to be persisted in the network
    - **Provider** - actor that offers his storage for Consumers to be used in monetary exchange
 - Entities
    - **Offer** - Offer created by Provider that announces availability of his storage space

>       - *Billing Period* - Specifies price for recurring period of time (e.g. 10 wei per day ) 

Reviewer: To me Billing Period only refers to a time frame. IT is not clear to me that the Price is included in that. An option could be BillingPlan or SubscriptionPlan or something like that probably. What do you think?

>       - *Billing Period* - Specifies price for recurring period of time (e.g. 10 wei per day ) 

Reviewer: To me  **BillingPeriod** only refers to a time frame. It is not clear to me at least that the Price is included in that concept. An option could be **BillingPlan** or **SubscriptionPlan** or something like that probably. What do you think?

Author: Good catch. Although in this instance it is indeed meant as timeframe, because inside the contracts the pricing mechanism is tracked as period-price tuple. I think it can be confusing because developer can think that `period` means the period for how long the Agreement is supposed to be valid, so I would call it "Billing Period" instead. 

With that I would then introduce "Billing Price" as well. 

We can introduce term "Billing Plan", yet I don't think it will be used inside the smartcontracts, but if you think that it is needed for the UI let me know.

Author: You are right the phrasing really suggested that it is more like a Billing Plan.

      - *Maximum Duration* - Duration which Consumer can maximally rent the storage space for 
      - *Total Capacity* - Total capacity that Provider is offering
      - *Free Capacity* - Capacity in given time that is left free for Consumers to use  
      - *Location* - Approximate location of the provider
      - *Offer Termination* - Provider can choose to discontinue his Offer. No new Agreements will be accepted, but current Agreements will be finished based on their current Billing Period.

>     - **Data Source** - Data Source is an abstract entity that defines the data that should be stored under an Agreement. There might be different mechanisms on how files should be retrieved, updated and removed.

Reviewer: We are calling it **StorageType** or **StorageSystem**, as it is something bigger than just a DataSource. Ideally I would use terminology you are handling in RIF Storage library which is "**Provider**", but the risk here is to get confused with the actual Provider of the Pinning service, so in that case would need to be renamed to **DataProvider** or something like that (Im not convinced)

Author: This is not a "Storage System" as you presented in the mockups. This is an abstraction on how to specify "what files to be pinned" and not "on which storage system they are stored". These two are linked as some Data Sources will be available only for some Storage Systems, but I would argue that these two things should stay separated.

I will add the "Storage System" term even though we won't use it for the first phase.

Reviewer 2: I am missing:

> *Swarm Feeds Source* - [That is also the future babe.](https://gph.is/1FD4aQ0)

      - *IPFS Source* - Hash pointer to immutable data source.
      - *Contract Source* - Pointer to different smart-contract that handles the mutability (eq. adding/removing files) 
      - *IPNS Source* - [That is the future babe.](https://gph.is/1FD4aQ0)

>     - **Agreement** - Consumer create Agreement for specific Provider's Offer and it specify Data Source Reference

Reviewer: ```suggestion
    - **Agreement** - Consumer creates Agreement for a specific Provider's Offer, specifying and the corresponding Data Source
```

      - **Data Reference** - Pointer to Data Source
      - **Agreement Termination** - Consumer can decide to terminate the agreement, which will take effect after finishing of current billing period.

>       - **Deposit Funds** - In order for Agreement to be active, it has to have available funds for each Billing Period renewal. Consumer can therefore deposit funds.

Reviewer: ```suggestion
      - **Deposit Funds** - In order for an Agreement to be active, it needs to have available funds for each Billing Period renewal. Consumer can therefore deposit funds.
```

>       - **Payout Funds** - Upon finishing of a Billing Period, the Provider can request the funds for finished period to be transferred to him. 

Reviewer: ```suggestion
      - **Payout Funds** - Upon finishing of a Billing Period, the Provider can request the funds for the finished period to be transferred to him. 
```

# RIF Marketplace Storage Pinning

```
npm i @rsksmart/rif-marketplace-storage-pinning
```

## PinningManager contract

### Message
The `StorageProvider` can emit an arbitrary message in the function calls `setStorageOffer` and `emitMessage`. 
The message is is initially intended to allow the `StorageProvider` to communicate his `nodeID`, but may be used for other purposes later.

The client should be instructed to enforce the following semantics:
- If `message` starts with `0x01`, then the following bits contain an arbitrary length `nodeID`.

## Glossary

 - Actors
    - **Consumer** - actor that wants his files to be persisted in the network
    - **Provider** - actor that offers his storage for Consumers to be used in monetary exchange
 - Entities
    - **Offer** - Offer created by Provider that announces availability of his storage space

>       - *Billing Periods* - Specifies periods that are available to used for billing   

Reviewer 2: I would say this is a broken English. I suggest:

> - *Billing Periods* - Specifies periods how often is payment required (e.g. every month, every 2 months...)

Author: Thanks @vojtechsimetka! Will fix, but for future, please save my work and create suggestions with: https://haacked.com/archive/2019/06/03/suggested-changes/ ;-)

>       - *Billing Prices* - Specifies prices for each Billing Period, resulting in period-price tuple.

Reviewer 3: Maybe add another glossary item:
*period-price tuple* - defines the Billing Price (per byte) which must be paid per Billing Period

Author: Well, I think that is redundant information...

>       - *Billing Prices* - Specifies prices for each Billing Period, resulting in period-price tuple.

Reviewer 3: Make it clear under this comment that the Billing Price is per Byte

      - *Maximum Duration* - Duration which Consumer can maximally rent the storage space for 
      - *Total Capacity* - Total capacity that Provider is offering

>       - *Free Capacity* - Capacity in given time that is left free for Consumers to use  

Reviewer 3: *Utilized Capacity* - Capacity (in bytes) which is currently being utilized

>       - *Free Capacity* - Capacity in given time that is left free for Consumers to use  

Reviewer 3: Specify per bytes. Maybe make clear the relationship between total capacity, free capacity and utilized capacity.

>       - *Location* - Approximate location of the provider

Reviewer 2: > - *Location* - Optional, self proclaimed location (country) from which the service is provided (where the data will be stored)

Reviewer 3: Was this a requirement for the mvp? 
If not, I would suggest we keep it out for now, to avoid confusion

>       - *Location* - Approximate location of the provider

Reviewer 2: > - *Location* - Optional, self proclaimed location (country) from which the service is provided (where the data will be stored)

>       - *Storage System* - Offer is always connected to one Storage System like IPFS or Swarm

Reviewer 3: Maybe Storage Network?

Author: Awesome! I like this term much more!

>       - *Offer Termination* - Provider can choose to discontinue his Offer. No new Agreements will be accepted, but current Agreements will be finished based on their current Billing Period.

Reviewer 3: > based on their current Billing Period.

could be misleading, as when a user has deposited multiple billing periods, the Offer continues for this time. 

Suggesting (...), but current Agreements will be finished based on the number of Billing Periods already purchased

    - **Data Source** - Data Source is an abstract entity that defines the data that should be stored under an Agreement. There might be different mechanisms on how files should be retrieved, updated and removed.
      - *IPFS Source* - Hash pointer to immutable data source.
      - *Contract Source* - Pointer to different smart-contract that handles the mutability (eq. adding/removing files) 
      - *IPNS Source* - [That is the future babe.](https://gph.is/1FD4aQ0)
      - *RNS Source* - [That is also the future babe.](https://gph.is/1FD4aQ0)
    - **Agreement** - Consumer creates Agreement for a specific Provider's Offer, specifying and the corresponding Data Source
      - *Data Reference* - Pointer to Data Source

>       - *Agreement Termination* - Consumer can decide to terminate the agreement, which will take effect after finishing of current billing period.

Reviewer 2: We should think about some form of graceful termination from the side of the provider as well

Author: Hmmm I guess, although not sure how to go about it. They should be notified about this termination but contacting them might not be as easy and reliable  as needed...

Reviewer 2: lets not bother with that now, it is ok as is

Reviewer 3: The provider can stop his offer by setting `capacity` to zero. This means, no new requests can be made and old requests can't be prolonged.

      - *Deposit Funds* - In order for an Agreement to be active, it needs to have available funds for each Billing Period renewal. Consumer can therefore deposit funds.

>       - *Payout Funds* - Upon finishing of a Billing Period, the Provider can request the funds for the finished periods to be transferred to him. 

Reviewer 2: Implementation detail, but we also need the consumer to withdraw funds as they may make mistake and overfund

Author: Okey, maybe "Refund Funds"?

Reviewer 2: Refund implies that the service was not satisfactory and the consumer gets the full amount out.

I think just *Withdraw  Deposit* could suffice

Reviewer 3: Consumer can at all times withdraw (but not from the current billing period)

Author: Yeah but for that he has to cancel the Agreement right? He can't just take out money without termination right?
@vojtechsimetka is this something required?

Reviewer 2: No he can always withdraw money until the amount that is locked which is the most imminent payment. I believe we already have this functionality

Author: Nope, currently only Provider can withdraw.

Author: Nope, currently only Provider can withdraw.

Reviewer 2: I guess that is required then

Reviewer: Yes I believe its not prepared to Withdraw as consumer unless you stop the Request. 
Lets add it here then we add it to the Contract.
